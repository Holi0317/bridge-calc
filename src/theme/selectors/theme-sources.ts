import i18next from "i18next";
import { themes } from "../color-presets";
import { IDropdownSource } from "../../material/dropdown";

/**
 * Available themes in dropdown source form.
 *
 * NOTE: This selector is not a usual selector. It does not require `state` argument
 * but a translate function argument.
 */
export function themeSourcesSelector(
  t: i18next.TFunction
): Array<IDropdownSource<string>> {
  return Array.from(themes.keys()).map(theme => ({
    value: theme,
    label: t(theme)
  }));
}
