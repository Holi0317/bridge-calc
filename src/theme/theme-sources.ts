import {themes} from './color-presets'
import {IDropdownSource} from '../types'

/**
 * Available themes in dropdown source form.
 *
 * This array is computed according to color-presets.ts upon import.
 * Therefore dynamic theme is not supported currently.
 */
export const themeSources: Array<IDropdownSource<string>> = Array.from(themes.keys())
  .map(theme => ({
    value: theme,
    label: theme
  }))
