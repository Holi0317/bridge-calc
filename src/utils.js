// @flow
/**
 * Generate a (hopefully) unique ID with 4 characters
 */
export function genID(): string {
  return Math.floor((1 + Math.random()) * 0x10000)
    .toString(16)
    .substring(1)
}

/**
 * NOTE: This is NOT lodash.nth method.
 * Get an element from given array.
 * If index is in range of array length, this would be identical to array[i].
 * If index is out of range, reduce it until it is in range of array and get element.
 * @example
 * nth(['a', 'b', 'c'], 4)
 * // -> 'b'
 * @param array - Array to be operated
 * @param index - Index of wanted element
 * @returns Element
 */
export function nth<T>(array: T[], index: number): T {
  if (index < 0) {
    throw new Error('Index should be a positive number')
  }
  if (!Number.isInteger(index)) {
    throw new Error('Index should be an integer')
  }
  return (index >= array.length) ? nth(array, index - array.length) : array[index]
}

/**
 * Sort array by putting frontIndex element to the first.
 * Does NOT mutate the original array.
 * @example
 * toFront(['a', 'b', 'c'], 1)
 * // -> ['b', 'c', 'a']
 * @param array - The original array
 * @param frontIndex - Index of element that need to be moved to front.
 * @returns
 */
export function toFront<T>(array: T[], frontIndex: number): T[] {
  if (frontIndex >= array.length) {
    throw new Error('frontIndex is out of boundary')
  }
  const newArray = []
  for (let i = frontIndex; i < array.length + frontIndex; i++) {
    newArray[i - frontIndex] = nth(array, i)
  }
  return newArray
}

/**
 * Test if there is error in the error object.
 * If no error, true will be returned. Vice versa for error object with error.
 * Meant to be used with validator.
 * @param error - Error object to be tested
 */
export function isOk(error: any): boolean {
  return JSON.stringify(error) === '{}'
}

/**
 * Remove any undefined value from the object given.
 * Does NOT mutate the original object. Instead, a shallow clone will be returned.
 * @param obj
 */
export function removeUndef(obj: any): any {
  const cloned = Object.assign({}, obj)
  Object.entries(cloned).forEach(([key, value]) => (value == null || value === '') && delete cloned[key])
  return cloned
}

/**
 * Fill object with value given if key does not exist or the original value is null.
 * Does NOT mutate the original object.
 * @example
 * const obj = {a: 'yo', b: 'bar', c: null}
 * const keys = ['a', 'b', 'c', 'd']
 * fill(obj, keys, 'filled')
 * // -> {a: 'yo', b: 'bar', c: 'filled', d: 'filled'}
 * @param obj - Object to be filled
 * @param keys - Keys need to be present in the return object
 * @param value - Value to be filled if filling action need to be done
 */
export function fill<T>(obj: {[key: string]: T}, keys: string[], value: T): {[key: string]: T} {
  const cloned = Object.assign({}, obj)
  for (const key of keys) {
    if (cloned[key] == null) {
      cloned[key] = value
    }
  }
  return cloned
}

/**
 * Typed Object.entries.
 * This is just a helper function to mock out flow's type checking
 */
export function toPairs(obj: any): any[] {
  return (Object: any).entries(obj)
}
