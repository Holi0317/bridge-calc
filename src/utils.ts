/**
 * Generate a (hopefully) unique ID with 4 characters
 */
export function genID() {
  return Math.floor((1 + Math.random()) * 0x10000)
    .toString(16)
    .substring(1);
}

/**
 * Generate an array form start to end.
 * Inspired by python's range method.
 *
 * @example
 * range(1, 10)
 * // -> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
 *
 * @param start Start from
 * @param end Ends with (inclusive)
 * @returns {number[]}
 */
export function range(start: number, end: number) {
  const result: number[] = [];
  for (let i = start; i <= end; i++) {
    result.push(i);
  }
  return result;
}

/**
 * NOTE: This is NOT lodash.nth method.
 * Get an element from given array.
 * If index is in range of array length, this would be identical to array[i].
 * If index is out of range, reduce it until it is in range of array and get element.
 * @example
 * nth(['a', 'b', 'c'], 4)
 * // -> 'b'
 * @param array - Array to be operated
 * @param index - Index of wanted element
 * @returns {T}
 */
export function nth<T>(array: T[], index: number): T {
  return (index > array.length) ? nth(array, index - array.length) : array[index];
}

/**
 * Sort array by putting frontIndex element to the first.
 * Does NOT mutate the original array.
 * @example
 * toFront(['a', 'b', 'c'], 1)
 * // -> ['b', 'c', 'a']
 * @param array
 * @param frontIndex
 * @returns {T}
 */
export function toFront<T>(array: T[], frontIndex: number): T[] {
  let newArray: T[] = [];
  for (let i = frontIndex; i < array.length + frontIndex; i++) {
    newArray[i-frontIndex] = nth(array, i);
  }
  return newArray;
}
