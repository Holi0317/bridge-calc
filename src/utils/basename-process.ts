export function stripTrailingSlash(basename: string): string {
  if (basename !== "/" && basename.substr(-1) === "/") {
    return basename.substr(0, basename.length - 1);
  }
  return basename;
}

export function stripHostname(basename: string): string {
  return new URL(basename).pathname;
}

export function basenameProcess(basename: string): string {
  return stripHostname(stripTrailingSlash(basename));
}
