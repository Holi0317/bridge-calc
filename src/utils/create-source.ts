import { DropdownSource } from "../material/dropdown";

/**
 * Create sources for react-toolbox's dropdown component.
 * The given array will be used as value in sources.
 * The label generated will be the stringfied version of value.
 * @param values - Array of value that will be source's value when generated.
 * @returns - Source array to be used in React-toolbox dropdown
 */
export function createSource(values: number[]): DropdownSource<number>[] {
  return values.map(s => ({ value: s, label: s + "" }));
}
