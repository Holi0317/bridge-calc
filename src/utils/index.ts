export {createSource} from './create-source'
export {dupe} from './dupe'
export {fillObj} from './fill-obj'
export {isOk} from './is-ok'
export {nth} from './nth'
export {removeUndef} from './remove-undef'
export {titleAugment} from './title-augment'
export {toFront} from './to-front'
export {toOrdinal} from './to-ordinal'
export {whitelistKeys} from './whitelist-keys'
export {isPlayerMapOf} from './is-player-map-of'
export {isArrayOf} from './is-array-of'
