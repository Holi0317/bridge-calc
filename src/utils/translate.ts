import i18next from "i18next";

/**
 * Mocked translate function for i18next.t
 * Used in testing only.
 *
 * @deprecated Use trans2 instead
 */
export function trans(key: string, options: any) {
  if (options == null) {
    return key;
  }
  let result = key;
  for (const k of Object.keys(options)) {
    const value = options[k];
    const template = `{{${k}}}`;
    result = result.replace(template, value);
  }
  return result;
}

/**
 * Arguments for translating using `i18next.t` function.
 *
 * Meant to be used in validation result.
 */
export interface ITranslateData {
  /**
   * Key of the translation.
   */
  key: string;

  /**
   * Optional options for the translation. Used for formatting.
   */
  options?: any;
}

/**
 * Convenient function for building ITranslateData.
 */
export function tData(key: string, options?: any): ITranslateData {
  return { key, options };
}

/**
 * Translate validation result.
 *
 * Meant to be called in component render. The TFunction should be obtained
 * by `react-i18next` package.
 *
 * TODO Rename this function to `trans` after migration is finished
 *
 * @param t Translate function from i18next
 * @param data Data to be translated
 * @returns formatted result
 */
export function trans2(t: i18next.TFunction, data: ITranslateData): string {
  return t(data.key, data.options);
}
