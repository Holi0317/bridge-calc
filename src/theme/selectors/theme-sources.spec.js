import { trans as t } from "../../utils/translate";
import { themeSourcesSelector } from "./theme-sources";

test("it should select translated drop down source", () => {
  const actual = themeSourcesSelector(t);
  expect(actual).toMatchSnapshot();
});
