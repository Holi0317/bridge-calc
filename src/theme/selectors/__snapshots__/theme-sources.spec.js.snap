// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should select translated drop down source 1`] = `
Array [
  Object {
    "label": "Pink",
    "value": "Pink",
  },
  Object {
    "label": "Indigo",
    "value": "Indigo",
  },
  Object {
    "label": "Light blue",
    "value": "Light blue",
  },
  Object {
    "label": "Blue Grey",
    "value": "Blue Grey",
  },
]
`;
